Non-Clifford Removal

nabe98 opened the pull request:
Implements non-Clifford removal method.
- [x] method to convert ZXGraphState into phase-gadget form (see Prop.4.16 in [1])
- [x] method to remove the YZ node with  a single neighbor (see Lem. 4.17)
- [x] method to merge YZ nodes with the same neighbors (see Lem. 4.18) 
- [x] repeat the above methods if these processes result in any new internal Clifford vertices (see Thm. 4.21)
- [x] add a demo in `examples/`
- [x] fix test_euler.py
- [x] fix local complementation and pivot
  - [x] replace `measurement_action` based impl. into `LocalClifford` based impl.
  - [x] update test_zxgraphstate 

Ref.
[1] M. Backens et al., Quantum 5, 421 (2021)

closes #8

masa10-f commented on tests/test_zxgraphstate.py:
```diff
@@ -1,15 +1,79 @@
+measurement_action_rc: dict[Plane, tuple[Plane, Callable[[float, float], float]]] = {
+    Plane.XY: (
+        Plane.XY,
+        lambda a_pi, alpha: (alpha if abs(a_pi % (2.0 * np.pi)) < atol else alpha + np.pi) % (2.0 * np.pi),
```
Does this work with a_pi = 2pi - epsilon?

masa10-f commented on tests/test_zxgraphstate.py:
```diff
@@ -1,15 +1,79 @@
+    from typing import Callable
+
+    Measurements = list[tuple[int, Plane, float]]
+
+measurement_action_lc_target: dict[Plane, tuple[Plane, Callable[[float], float]]] = {
+    Plane.XY: (Plane.XZ, lambda angle: angle + np.pi / 2),
+    Plane.XZ: (Plane.XY, lambda angle: -angle + np.pi / 2),
+    Plane.YZ: (Plane.YZ, lambda angle: angle + np.pi / 2),
+}
+measurement_action_lc_neighbors: dict[Plane, tuple[Plane, Callable[[float], float]]] = {
+    Plane.XY: (Plane.XY, lambda angle: angle + np.pi / 2),
+    Plane.XZ: (Plane.YZ, lambda angle: angle),
+    Plane.YZ: (Plane.XZ, operator.neg),
+}
+measurement_action_pv_target: dict[Plane, tuple[Plane, Callable[[float], float]]] = {
+    Plane.XY: (Plane.YZ, operator.neg),
+    Plane.XZ: (Plane.XZ, lambda angle: (np.pi / 2 - angle)),
+    Plane.YZ: (Plane.XY, operator.neg),
+}
+measurement_action_pv_neighbors: dict[Plane, tuple[Plane, Callable[[float], float]]] = {
+    Plane.XY: (Plane.XY, lambda angle: (angle + np.pi) % (2.0 * np.pi)),
+    Plane.XZ: (Plane.XZ, lambda angle: -angle % (2.0 * np.pi)),
+    Plane.YZ: (Plane.YZ, lambda angle: -angle % (2.0 * np.pi)),
+}
+atol = 1e-9
+measurement_action_rc: dict[Plane, tuple[Plane, Callable[[float, float], float]]] = {
+    Plane.XY: (
+        Plane.XY,
+        lambda a_pi, alpha: (alpha if abs(a_pi % (2.0 * np.pi)) < atol else alpha + np.pi) % (2.0 * np.pi),
+    ),
+    Plane.XZ: (
+        Plane.XZ,
+        lambda a_pi, alpha: (alpha if abs(a_pi % (2.0 * np.pi)) < atol else -alpha) % (2.0 * np.pi),
+    ),
+    Plane.YZ: (
+        Plane.YZ,
+        lambda a_pi, alpha: (alpha if abs(a_pi % (2.0 * np.pi)) < atol else -alpha) % (2.0 * np.pi),
+    ),
+}
```
Please leave comments about what `lc`, `pv`, and `rc` are here. If possible, please cite the source

masa10-f commented on tests/test_zxgraphstate.py:
```diff
@@ -1,15 +1,79 @@
+if TYPE_CHECKING:
+    from typing import Callable
+
+    Measurements = list[tuple[int, Plane, float]]
```
Why not use the `MeasBasis` class?

masa10-f commented on tests/test_zxgraphstate.py:
```diff
@@ -102,166 +166,118 @@ def test_local_complement_fails_with_input_node(zx_graph: ZXGraphState) -> None:
 
-def test_local_complement_with_no_edge(zx_graph: ZXGraphState) -> None:
-    """Test local complement with a graph with no edge."""
+@pytest.mark.parametrize("plane", [Plane.XY, Plane.XZ, Plane.YZ])
```
`list[Plane]` is simpler

masa10-f replied:
Sorry for confusing you. I meant replacing `[Plane.XY, Plane.XZ, Plane.YZ]` with `list(Plane)` is simpler in the parameterized decorator.

masa10-f reviewed (commented):
Nice work! I left several comments and questions about the performance and logic. Please check them.

nabe98 commented on graphix_zx/zxgraphstate.py:
```diff
@@ -509,3 +474,124 @@ def remove_cliffords(self, atol: float = 1e-9) -> None:
+            self.merge_yz_to_xy()
+            self.merge_yz_nodes()
+            if not any(
+                is_clifford_angle(self.meas_bases[node].angle, atol) and self.is_removable_clifford(node, atol)
+                for node in self.physical_nodes - self.input_nodes - self.output_nodes
+            ):
```
For this part, 
only `self.is_removable_clifford(node, atol)` is required (the former condition is unnecessary).

nabe98 commented on graphix_zx/random_objects.py:
```diff
@@ -82,3 +83,50 @@ def get_random_flow_graph(
+    edge_p: float = 0.5,
+    angle_list: list | None = None,
+) -> MBQCCircuit:
+    """Generate a random MBQC circuit which has gflow.
```
I forgot to delete "which has gflow".
The current implementation does not guarantee the existence of gflow.

nabe98 commented on graphix_zx/zxgraphstate.py:
```diff
@@ -509,3 +474,124 @@ def remove_cliffords(self, atol: float = 1e-9) -> None:
+
+                    new_angle = (self.meas_bases[u].angle + self.meas_bases[v].angle) % (2.0 * np.pi)
+                    self.set_meas_basis(u, PlannerMeasBasis(Plane.YZ, new_angle))
+                    self.remove_physical_node(v)
```
Thanks for the suggestion! I’ve incorporated your feedback. Please take a look.

nabe98 commented:
@masato-fuk 
Thanks for the review! I've fixed all the issues you pointed out.

As for the tests, there are no problems with the current ones, but in order to fully cover all the conditional branches implemented in `zxgraphstate.py`, three additional tests are needed. (I already have drafts for them, but the calculations still need to be revised.)
These additional tests are planned to be included in a subsequent PR.

masa10-f commented on graphix_zx/zxgraphstate.py:
```diff
@@ -509,3 +485,116 @@ def remove_cliffords(self, atol: float = 1e-9) -> None:
+            A pair of adjacent nodes that are both measured in the YZ-plane, or None if no such pair exists.
+        """
+        yz_nodes = {node for node, basis in self.meas_bases.items() if basis.plane == Plane.YZ}
+        for u in yz_nodes:
+            for v in self.get_neighbors(u):
+                if v in yz_nodes:
+                    return (min(u, v), max(u, v))
```
The double-for-loop is not preferable from a performance perspective. Also, is it possible to cache the result and update outside the method?

nabe98 replied:
Thanks for pointing out! I have two comments about this.

As for double-for-loop issue, the following would be better.
```python
for u, v in self.physical_edges:
    if u in yz_nodes and v in yz_nodes:
        return (min(u, v), max(u, v))
```

> is it possible to cache the result and update outside the method?

Just to clarify—are you suggesting caching the first found YZ-adjacent pair?
I’m wondering whether that would be worthwhile in this case. Since pivot and local_complement update the measurement planes inside `convert_to_phase_gadget`, any cached pair might become invalid after each step.
If you have a specific use case in mind where caching helps, I’d be curious to hear more!

masa10-f replied:
No, I'm considering a way to find all the YZ edges and update them every local complementation and pivot call. Calling this method may not be efficient, I guess. Please check the overall performance via performance profiler or something, and compare it with the theoretical calculation complexity derived in the reference paper.

masa10-f replied:
I don't understand why these operations incur O(N) operations, as we can know which nodes are affected in the LC and edge updates. Anyway, it's not the emergent, so you can address it in another issue.

masa10-f commented on graphix_zx/zxgraphstate.py:
```diff
@@ -509,3 +485,116 @@ def remove_cliffords(self, atol: float = 1e-9) -> None:
+        for neighbors, nodes in neighbor_groups.items():
+            if len(nodes) < min_nodes or len(neighbors) < min_nodes:
+                continue
+            u = nodes[0]
+            for v in nodes[1:]:
+                new_angle = (self.meas_bases[u].angle + self.meas_bases[v].angle) % (2.0 * np.pi)
+                self.set_meas_basis(u, PlannerMeasBasis(Plane.YZ, new_angle))
+                self.remove_physical_node(v)
```
Can't we merge all the nodes into one?

nabe98 replied:
How about this? This is equivalent and faster.
```python
new_angle = sum(self.meas_bases[v].angle for v in nodes) % (2.0 * np.pi)
self.set_meas_basis(nodes[0], PlannerMeasBasis(Plane.YZ, new_angle))
for v in nodes[1:]:
    self.remove_physical_node(v)
```

masa10-f replied:
LGTM

masa10-f commented on graphix_zx/zxgraphstate.py:
```diff
@@ -509,3 +485,116 @@ def remove_cliffords(self, atol: float = 1e-9) -> None:
+                self.remove_physical_node(v)
+
+    def full_reduce(self, atol: float = 1e-9) -> None:
+        """Reduce removable non-Clifford vertices from the graph state.
```
Only non-Clifford nodes?

nabe98 replied:
No, also removes all Clifford nodes. I updated this in the new commits.

masa10-f reviewed (commented):
Please check my comments

`_update_node_measurement` seems not to be used.

masa10-f commented on graphix_zx/random_objects.py:
```diff
@@ -2,6 +2,7 @@
 
 This module provides:
 - get_random_flow_graph: Generate a random flow graph.
+- random_circ: Generate a random MBQC circuit with gflow.
```
```suggestion
- random_circ: Generate a random MBQC circuit.
```

masa10-f commented:
@nabe98 
The ZXGraphState class has too many private methods. Can you simplify them? We have to update gflow inside the routines afterwards, so the simple implementation is crucial.

nabe98 commented on examples/measurement_pattern_simplification.py:
```diff
@@ -0,0 +1,42 @@
+"""Basic example of simplifying a measurement pattern via a ZX-diagram simplification.
```
No, this comment was from a previous demo code, so I will update the docstrings as follows.

```txt
Basic example of simplifying a ZX-diagram.

By using the `full_reduce` method,
we can remove all the internal Clifford nodes and some non-Clifford nodes from the graph state,
which generates a simpler ZX-diagram.
This example is a simple demonstration of the simplification process.
```

nabe98 commented:
@masato-f
Thanks for the review!
The main changes are as follows:
1. Removed some private methods to improve readability
2. Simplified the preprocessing logic for remove_clifford

masa10-f commented on graphix_zx/random_objects.py:
```diff
@@ -82,3 +86,48 @@ def get_random_flow_graph(
+    edge_p : float, optional
+        probability of adding CZ gate, by default 0.5
+    angle_candidates : collections.abc.Sequence[float], optional
+        list of angles, by default (0, np.pi / 3, 2 * np.pi / 3, np.pi)
```
```suggestion
        sequence of angles, by default (0, np.pi / 3, 2 * np.pi / 3, np.pi)
```

masa10-f commented on graphix_zx/zxgraphstate.py:
```diff
@@ -30,18 +32,35 @@ class ZXGraphState(GraphState):
         qubit indices
     local_cliffords : dict[int, LocalClifford]
         local clifford operators
+    _clifford_rules : tuple[tuple[Callable[[int, float], bool], Callable[[int], None]], ...]
```
No need to write private attributes inside the docstring.

masa10-f commented on graphix_zx/zxgraphstate.py:
```diff
@@ -30,18 +32,35 @@ class ZXGraphState(GraphState):
     def __init__(self) -> None:
         super().__init__()
 
+    @cached_property
+    def _clifford_rules(self) -> tuple[tuple[Callable[[int, float], bool], Callable[[int], None]], ...]:
+        """List of rules (check_func, action_func) for removing local clifford nodes.
+
+        The rules are applied in the order they are defined.
+        """
+        return (
+            (self._needs_lc, self.local_complement),
+            (self._needs_nop, lambda _: None),
+            (
+                self._needs_pivot,
+                lambda node: self.pivot(node, min(self.get_neighbors(node) - self.input_nodes)),
+            ),
+        )
```
Why cached_property?

nabe98 replied:
I initially used this to reduce overhead from repeated construction. However, since the logic internally depends on the dynamic state of `self` (e.g. get_neighbors, input_nodes), caching can lead to incorrect behavior as the graph evolves. I’ll change it to @property to ensure it always reflects the latest state.
